BeastH7 v2 support

e-fominov opened the pull request:
I was trying to run existing firmware on the latest BeastH7 v2 board and it does not work because IMU was changed to BMI270 and now the board has no barometer.
This PR is inspired by latest BeastF7 v2 hwdef and allows building a compatible binary for the latest board.
Latest product page: https://shop.iflight-rc.com/Beast-H7-V2-55A-AIO-MPU6000-25.5-25.5-pro1588

I have compiled and successfully run this code on my boards (H7 v2). IMU works well without warning/error messages in Mission Planner.

khancyr reviewed (changes requested):
Replacing the previous hwdef should only be done if the previous boards isn't sold anymore.
But even without I think it will be better to create a new config as the no baro change isn't straight forward.

About the commit, you could you change you commits to start with AP_HAL_Chibios:

andyp1per commented:
Ah ok, this is for the H7. Please look at my changes for the F7 and do something similar - it's possible to support both boards in the same hwdef

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/BeastH7/README.md:
```diff
@@ -1,16 +1,16 @@
 # iFlight Beast H7 55A AIO Flight Controller
 
-https://shop.iflight-rc.com/index.php?route=product/product&product_id=1376
+https://shop.iflight-rc.com/Beast-H7-V2-55A-AIO-MPU6000-25.5-25.5-pro1588
```
Please provide links to both v1 and v2

macourteau commented:
I'm working on an Arducopter quad that would use this FC (Beast V2), so thanks for working on the hwdef changes needed to support it!

It seems like there was no way to get a compass hooked up with v1 of the Beast FC, but I see on the v2 of the board that RX3/TX3 are now both broken out on the PCB. If I'm reading the H743 datasheet right, it looks like these would be PB11/PB10, which do support I2C (!), so it might be possible to hook up the compass from e.g. an M8Q-5883 there.

If this is correct, would there be any changes needed to the hwdef to allow using that I2C bus? (apologies if this obvious, I'm new to the ardupilot codebase)

For reference: comparing the wiring diagrams between v1 ([here](https://shop.iflight-rc.com/image/catalog/download/SucceX/Beast-F7-AIO-45A-Wiring-Diagram.png)) and v2 ([here](https://shop.iflight-rc.com/image/catalog/product/Beast-H7-V2/Beast-H743-55A-AIO-V2.png)), it seems like v2 removed TX2 and added TX3. 

It also seems like the beeper pin may have [changed](https://github.com/betaflight/betaflight/blob/master/src/main/target/IFLIGHT_H743_AIO_V2/target.h#L29) from the [original one](https://github.com/betaflight/betaflight/blob/master/src/main/target/IFLIGHT_H743_AIO/target.h#L29)?

e-fominov commented:
@macourteau  thanks for your links, it really looks like we can get I2C here. Also, the TX2 pin is still there - on the back side of the board near DJI connector pins. I have some i2c compass, will test and update PR

@khancyr now i see more differences between V2 and V1, this is definitely a new board. I will update PR to add a new board

andyp1per commented:
> It seems like there was no way to get a compass hooked up with v1 of the Beast FC, but I see on the v2 of the board that RX3/TX3 are now both broken out on the PCB. If I'm reading the H743 datasheet right, it looks like these would be PB11/PB10, which do support I2C (!), so it might be possible to hook up the compass from e.g. an M8Q-5883 there.

What a great catch! Yes you are right. The same is true on the F7, so the current definition is wrong for the F7 v2 as well. @e-fominov do you feel like creating two new boards as part of your PR?

e-fominov commented:
@andyp1per I have added a new board BeastH7v2 and reverted my previous commits. The bootloader is copy-pasted binary from BeastH7.
Regarding the F7v2 board - I see that it also has RX3/TX3 pins that could be remapped into I2C. But doing such changes without a board for testing looks brittle for me.

@macourteau, I have tested the compass (M8Q-5883), the I2C seems working. RX3/TX3 are now used for I2C, the README.md file contains wiring diagram. RX2/TX2 also tested, seems working. BZ- pin changed but not tested.

andyp1per reviewed (changes requested):
Please do the new hwdef by inclusion and undef/redefining of the pins. Take a look at KakuteF4MIni for an example. You need to do this for both the bootloader and the regular hwdef

e-fominov commented:
@andyp1per updated the hwdef to use include. How can I test building the bootloader?

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/BeastH7/hwdef.dat:
```diff
@@ -6,7 +6,7 @@
 MCU STM32H7xx STM32H743xx
 
 # board ID for firmware load
-APJ_BOARD_ID 1025
```
Need to keep the old id

e-fominov replied:
fixed

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/BeastH7v2/hwdef-bl.dat:
```diff
@@ -0,0 +1 @@
+include ../BeastH7/hwdef-bl.dat
```
Need to add the new board id

e-fominov replied:
board id added

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/BeastH7v2/hwdef.dat:
```diff
@@ -0,0 +1,42 @@
+
+# I2C2 for compass. These pins can also be used as USART3
+PB10 I2C2_SCL I2C2 
+PB11 I2C2_SDA I2C2 
```
You should add an ALT configuration to allow these to be used as a UART. Tridge thinks this should work - just put ALT(1) on the end of the alternate pin definitions

e-fominov replied:
fixed, but not tested with real hardware yet

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/BeastH7v2/hwdef.dat:
```diff
@@ -0,0 +1,42 @@
+
+# one IMU
+IMU BMI270 SPI:bmi270 ROTATION_ROLL_180_YAW_225
+define HAL_DEFAULT_INS_FAST_SAMPLE 1
```
fast sampling has no meaning on the BMI270

e-fominov replied:
undefined fast sampling

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/BeastH7v2/hwdef.dat:
```diff
@@ -0,0 +1,42 @@
+include ../BeastH7/hwdef.dat
+undef APJ_BOARD_ID
+undef PD15 PB10 PB11 
+undef HAL_SERIAL3_PROTOCOL
```
You need to define serial2 protocol as a default to 23 since that is defined as the default serial RX uart. You also need to disable the usart2 TX pin since that is not available.

e-fominov replied:
@andyp1per , TX2 exists on the back side of board. It also has printed marks on it. I have tested RX2/TX2 with connecting UART-USB converter and using Mavlink2 protocol with Mission Planner. I have updated the pictures of the pinout to show that. Serial2 with default 23 made in today's commit

andyp1per commented:
> How can I test building the bootloader?

./Tools/scripts/build_bootloaders.py BeastH7v2 ?

e-fominov commented:
Work in progress. Got the first take off today. Need motor pin changes. Firmware is not ready to be used now.

andyp1per commented:
Well I'm glad the BMI270 driver appears to work - I have not flown that as yet!

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/BeastH7v2/hwdef.dat:
```diff
@@ -0,0 +1,46 @@
+APJ_BOARD_ID 1056
+
+# V2 has different motor mapping
+PB4 TIM3_CH1 TIM3 PWM(2) GPIO(51)     # 3
```
You need to add a BIDIR to one one these. Bi-dir applies in channel pairs - 1/2 or 3/4

e-fominov replied:
fixed

macourteau commented on libraries/AP_HAL_ChibiOS/hwdef/BeastH7v2/README.md:
```diff
@@ -0,0 +1,87 @@
+## Features
+
+ - MCU: BGA-STM32H743
+ - Gyro: BMI2700
```
```suggestion
 - Gyro: BMI270
```

e-fominov replied:
fixed

macourteau commented:
Question: would it make sense to have the RC input configured on USART1 or UART4 instead of USART2? That way, for receivers that use both RX and TX (e.g. CRSF receivers), both wires would be either soldered to the board (UART4) or on the DJI connector (USART1), but not a mix of both (USART2)?

andyp1per commented:
> Question: would it make sense to have the RC input configured on USART1 or UART4 instead of USART2? That way, for receivers that use both RX and TX (e.g. CRSF receivers), both wires would be either soldered to the board (UART4) or on the DJI connector (USART1), but not a mix of both (USART2)?

The user can obviously configure whatever they like - its just a case of what makes the most sense given the board documentation

andyp1per commented:
I'll have my F7v2 PR up later today - there are a couple of other changes that you need

e-fominov commented:
@andyp1per When board is idle and laying on the table, EKF gives non-zero ground speed(accelerating and stops growing near 700 m/s) and it thinks that drone is rotating (non-zero roll/pitch speed), while it really lays on the table (accelerometer calibrated). IMU noise is not high, gyro has zero mean, accelerometer has non-zero mean, but that may be fine. Is it correct to have non-zero ground speed and non-zero roll/pitch speed here? May be it is because no GPS connected? Tested with two boards.

![imu-1](https://user-images.githubusercontent.com/13793158/148102430-efb017d1-68a3-41fa-9e94-53fd67b90a19.jpg)
![gyro-noise](https://user-images.githubusercontent.com/13793158/148102444-58106468-8b2c-4d5b-98df-8a4792665808.jpg)
![acc-noise](https://user-images.githubusercontent.com/13793158/148102448-79bea1b7-64df-43dc-b914-5839aca35914.jpg)

[Board idle laying on the table log](https://drive.google.com/file/d/1PN4mWCd_Eg9C3x_0wmUdUxWvC2GUxCp2/view?usp=sharing)
[Parameters](https://drive.google.com/file/d/1b1NYJjlfZg8sJcX0YwqV5AHMu8oFXg_y/view?usp=sharing)

andyp1per commented:
I see something similar, but the fact that there is no GPS may be the culprit, not sure

andyp1per commented:
This is my PR. https://github.com/ArduPilot/ardupilot/pull/19676
Seems to work, all the motors turn and I get telemetry with bdshot.
Does the neopixel work for you? It's not working for me, but unless they have published the wrong pin I can't see what the issue is

andyp1per commented:
My PR has been merged

e-fominov commented:
@andyp1per I have rebased the commits to fix the conflict, undefined I2C1. Have no chance to test NeoPixel yet, but will try later

rmackay9 commented:
@andyp1per,

I think you mentioned that BeastH7v2 support was in 4.2 but I see this PR.  Maybe an alternative to this PR has been merged?

andyp1per commented:
@rmackay9 - arggh, we should try and get this merged!
